File: governance-service/src/main/java/com/msval/governance/cdm/Quantity.java
package com.msval.governance.cdm;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * F3 — Kubernetes quantity grammar, canonicalization, and quantity expressions.
 * Canonical integers: cpu → millicores, memory → bytes.
 * Behavioural mirror of msval/core/cdm/quantity.py; TEST-002 pins the edges.
 */
public final class Quantity {

    private static final Pattern QTY_RE =
            Pattern.compile("^([0-9]+(?:\\.[0-9]+)?)(m|Ki|Mi|Gi|Ti|k|M|G|T)?$");

    private static final Map<String, Double> MEM_FACTORS = Map.of(
            "Ki", Math.pow(1024, 1), "Mi", Math.pow(1024, 2), "Gi", Math.pow(1024, 3), "Ti", Math.pow(1024, 4),
            "k", 1000d, "M", Math.pow(1000, 2), "G", Math.pow(1000, 3), "T", Math.pow(1000, 4));

    private static final Pattern EXPR_TOKEN = Pattern.compile("\\s*([+-])\\s*");

    private Quantity() {
    }

    /** Parse a quantity string to canonical long (cpu: millicores, memory: bytes). */
    public static long parse(String qty, String dimension) {
        if (!"cpu".equals(dimension) && !"memory".equals(dimension)) {
            throw new QuantityException("unknown dimension '" + dimension + "'");
        }
        Matcher m = QTY_RE.matcher(qty.strip());
        if (!m.matches()) {
            throw new QuantityException("invalid quantity '" + qty + "'");
        }
        double num = Double.parseDouble(m.group(1));
        String unit = m.group(2);
        double value;
        if ("cpu".equals(dimension)) {
            if ("m".equals(unit)) {
                value = num;
            } else if (unit == null) {
                value = num * 1000;
            } else {
                throw new QuantityException("unit '" + unit + "' is not a cpu unit in '" + qty + "'");
            }
        } else {
            if ("m".equals(unit)) {
                throw new QuantityException("'m' is not a memory unit in '" + qty + "'");
            }
            value = num * (unit == null ? 1d : MEM_FACTORS.get(unit));
        }
        if (value < 0) {
            throw new QuantityException("negative quantity '" + qty + "'");
        }
        // Python round() is half-to-even; Math.rint matches that (Math.round would be half-up).
        return (long) Math.rint(value);
    }

    /** F3: qty (('+'|'-') qty)* — same-dimension terms only, canonical long result. */
    public static long parseExpr(String expr, String dimension) {
        String s = expr.strip();
        // Mirror re.split(r"\s*([+-])\s*", s): parts alternate term, op, term, ...
        List<String> parts = new ArrayList<>();
        Matcher m = EXPR_TOKEN.matcher(s);
        int last = 0;
        while (m.find()) {
            parts.add(s.substring(last, m.start()));
            parts.add(m.group(1));
            last = m.end();
        }
        parts.add(s.substring(last));
        if (parts.isEmpty() || parts.get(0).isEmpty()) {
            throw new QuantityException("empty quantity expression '" + expr + "'");
        }
        long total = parse(parts.get(0), dimension);
        for (int i = 1; i < parts.size(); i += 2) {
            String op = parts.get(i);
            String term = i + 1 < parts.size() ? parts.get(i + 1) : null;
            if (term == null) {
                throw new QuantityException("dangling operator in '" + expr + "'");
            }
            total += "+".equals(op) ? parse(term, dimension) : -parse(term, dimension);
        }
        return total;
    }
}

File: src/msval/core/cdm/quantity.py
"""F3 — Kubernetes quantity grammar, canonicalization, and quantity expressions.

Canonical integers: cpu -> millicores, memory -> bytes. Both engines implement
identically; TEST-002 pins the edges.
"""
from __future__ import annotations

import re

_QTY_RE = re.compile(r"^(?P<num>[0-9]+(?:\.[0-9]+)?)(?P<unit>m|Ki|Mi|Gi|Ti|k|M|G|T)?$")

_MEM_FACTORS = {"Ki": 1024, "Mi": 1024**2, "Gi": 1024**3, "Ti": 1024**4,
                "k": 1000, "M": 1000**2, "G": 1000**3, "T": 1000**4, None: 1}


class QuantityError(ValueError):
    """Invalid quantity string or dimension mismatch."""


def parse(qty: str, dimension: str) -> int:
    """Parse a quantity string to canonical int (cpu: millicores, memory: bytes)."""
    if dimension not in ("cpu", "memory"):
        raise QuantityError(f"unknown dimension {dimension!r}")
    m = _QTY_RE.match(qty.strip())
    if not m:
        raise QuantityError(f"invalid quantity {qty!r}")
    num, unit = float(m.group("num")), m.group("unit")
    if dimension == "cpu":
        if unit == "m":
            value = num
        elif unit is None:
            value = num * 1000
        else:
            raise QuantityError(f"unit {unit!r} is not a cpu unit in {qty!r}")
    else:
        if unit == "m":
            raise QuantityError(f"'m' is not a memory unit in {qty!r}")
        value = num * _MEM_FACTORS[unit]
    if value < 0:
        raise QuantityError(f"negative quantity {qty!r}")
    return round(value)


_EXPR_TOKEN = re.compile(r"\s*([+-])\s*")


def parse_expr(expr: str, dimension: str) -> int:
    """F3: qty (('+'|'-') qty)* — same-dimension terms only, canonical int result."""
    parts = _EXPR_TOKEN.split(expr.strip())
    if not parts or not parts[0]:
        raise QuantityError(f"empty quantity expression {expr!r}")
    total = parse(parts[0], dimension)
    for i in range(1, len(parts), 2):
        op, term = parts[i], parts[i + 1] if i + 1 < len(parts) else None
        if term is None:
            raise QuantityError(f"dangling operator in {expr!r}")
        total += parse(term, dimension) if op == "+" else -parse(term, dimension)
    return total
